contracts/games/session-activity/SessionActivityDeployer.sol

```diff
@@ -0,0 +1,90 @@
+// Copyright (c) Immutable Pty Ltd 2018 - 2024
+// SPDX-License-Identifier: Apache 2
+// solhint-disable not-rely-on-time
+
+pragma solidity ^0.8.19;
+
+import {AccessControlEnumerable} from "@openzeppelin/contracts/access/AccessControlEnumerable.sol";
+import {SessionActivity} from "./SessionActivity.sol";
+
+error Unauthorized();
+error NameAlreadyRegistered();
+
+/**
+ * @title SessionActivityDeployer - A factory contract that deploys SessionActivity contracts and tracks their addresses and names
+ * @author Immutable
+ * @dev The SessionActivityDeployer contract is not designed to be upgradeable or extended.
+ */
+contract SessionActivityDeployer is AccessControlEnumerable {
+    /// @notice Indicates that an account has registered session activity
+    event SessionActivityDeployed(address indexed account, string indexed name);
+
+    /// @notice Mapping of deployed SessionActivity contract addresses to their names
+    /// @dev To get a list of all deployed contract names, iterate over the deployedContracts array and use this mapping
+    mapping(address deployedContract => string name) public sessionActivityNames;
+
+    /// @notice Mapping of SessionActivity contract names to their addresses
+    /// @dev To get a list of all deployed contract addresses, iterate over the names array and use this mapping
+    mapping(string name => address deployedContract) public sessionActivityContracts;
+
+    /// @notice Array of deployed SessionActivity contracts
+    address[] public deployedContracts;
```

Comment:
This declaration will result in an accessor that returns an array. If there are a lot of values, this could result in a out of gas error (yes, for a view call)

Reply:
That is, if we get to 100,000 (or maybe just 10,000) values. If we feel that it is unlikely to ever happen, then it is probably OK. An alternative would be to have a function to return how many contracts have been deployed, and another function to request, starting at contract N, return information for up to M contracts.